src/libtools/reducedgraph.cc

```diff
@@ -0,0 +1,208 @@
+    }  // Chains same size
+  }
+  return false;
+}
+
+void ReducedGraph::init_(vector<ReducedEdge> reduced_edges,
+                         unordered_map<int, GraphNode> nodes) {
+  vector<Edge> edges;
+  nodes_ = nodes;
+  for (const ReducedEdge& reduced_edge : reduced_edges) {
+    Edge edge(reduced_edge.getEndPoint1(), reduced_edge.getEndPoint2());
+    edges.push_back(edge);
+    if (expanded_edges_.count(edge)) {
+      bool match = compareChainWithChains_(reduced_edge.getChain(),
+                                           expanded_edges_[edge]);
+      if (!match) {
+        expanded_edges_[edge].push_back(reduced_edge.getChain());
+      }
+    } else {
+      expanded_edges_[edge].push_back(reduced_edge.getChain());
+    }
+  }
+  edge_container_ = EdgeContainer(edges);
+  calcId_();
+}
+
+set<int> getAllVertices_(const std::vector<ReducedEdge>& reduced_edges) {
+  set<int> vertices;
+  for (auto reduced_edge : reduced_edges) {
+    vector<int> chain = reduced_edge.getChain();
+    for (const int vertex : chain) {
+      vertices.insert(vertex);
+    }
+  }
+  return vertices;
+}
+
+ReducedGraph::ReducedGraph(std::vector<ReducedEdge> reduced_edges) {
+
+  set<int> vertices = getAllVertices_(reduced_edges);
+  unordered_map<int, GraphNode> nodes;
+  for (const int vertex : vertices) {
+    GraphNode gn;
+    nodes[vertex] = gn;
+  }
+  init_(reduced_edges, nodes);
+}
+
+ReducedGraph::ReducedGraph(std::vector<ReducedEdge> reduced_edges,
+                           unordered_map<int, GraphNode> nodes) {
+  set<int> vertices = getAllVertices_(reduced_edges);
+  if (nodes.size() < vertices.size()) {
+    throw invalid_argument(
+        "The number of nodes passed into a reduced graph "
+        "must be greater or equivalent to the number of vertices");
+  }
+  for (const int vertex : vertices) {
+    if (nodes.count(vertex) == 0) {
+      throw invalid_argument("A vertex is missing its corresponding node.");
+    }
+  }
+  init_(reduced_edges, nodes);
+}
+
+Graph ReducedGraph::expandGraph() {
+  vector<Edge> all_expanded_edges;
+  for (const pair<Edge, vector<vector<int>>> edge_and_vertices :
+       expanded_edges_) {
+    vector<vector<Edge>> edges_local = expandEdge(edge_and_vertices.first);
+    for (vector<Edge>& edges : edges_local) {
+      all_expanded_edges.insert(all_expanded_edges.end(), edges.begin(),
+                                edges.end());
+    }
+  }
+  return Graph(all_expanded_edges, nodes_);
+}
+
+vector<vector<Edge>> ReducedGraph::expandEdge(Edge edge) {
+  assert(expanded_edges_.count(edge));
+  vector<vector<Edge>> all_edges;
+  vector<vector<int>> chains = expanded_edges_[edge];
+  for (vector<int> vertices : chains) {
+    vector<Edge> edges;
+    for (size_t index = 0; index < (vertices.size() - 1); ++index) {
+      Edge edge(vertices.at(index), vertices.at(index + 1));
+      edges.push_back(edge);
+    }
+    all_edges.push_back(edges);
+  }
+  return all_edges;
+}
+
+set<int> getAllConnectedVertices_(
```

Comment:
const?

Reply:
This is not a member method it is a private internal function I cannot use const.

Reply:
what do you mean by that?

Reply:
Where did you want me to put the const?